### PR TITLE
Limit maximum tooltip width to prevent overflow

### PR DESCRIPTION
**Pull request by normalid-awa**
Replace the default `IHasTooltip` with `IHasCustomTooltip<LocalisableString>` and use the new custom tooltip `CommentTooltip` with `TextFlowContainer` to limit the width of the tooltip.

| Master | image | image |
| ---------| - | - |
| PR | image | image |

### CODE REVIEW

**Comment by smoogipoo** [comment]
Agree with the above, should be fairly easy to make it use TFC.

**Comment by normalid-awa** [comment]
I've changed the `OsuTooltip` text to TFC and limited the width, so it should be fixing all potential visual bugs (tooltip overflow).

File: osu.Game/Graphics/Cursor/OsuTooltipContainer.cs [hunk 1]
```diff
@@ -65,10 +72,16 @@ public OsuTooltip()
                         RelativeSizeAxes = Axes.Both,
                         Alpha = 0.9f,
                     },
-                    text = new OsuSpriteText
+                    text = new TextFlowContainer(f =>
+                    {
+                        f.Font = OsuFont.GetFont(weight: FontWeight.Regular);
+                        f.Truncate = true;
+                        f.MaxWidth = max_width;
```
Comment (smoogipoo) [inline osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:79]: This doesn't look good at all and doesn't make sense in the context of how other apps handle this.

`AllowMultiline` can't be used because it crashes on retrieving the base line height (not allowed on multiline `SpriteText` for some reason).

But, even if multiline was to be used, TFC will still not handle correctly and will put the following word always on a new line because the spritetext covers the entire horizontal area.

Basically, TFC needs a lot of work (framework-side) to do things correctly but in the meantime I'd rather not have this here because it looks wrong.

File: osu.Game/Graphics/Cursor/OsuTooltipContainer.cs [hunk 2]
```diff
@@ -27,13 +27,18 @@ public OsuTooltipContainer(CursorContainer cursor)
 
         public partial class OsuTooltip : Tooltip
         {
+            private const float max_width = 1024;
```
Comment (smoogipoo) [inline osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:30]: I think this is a too long. 500 is more in line with what I'd expect going by browsers.

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Graphics/Cursor/OsuTooltipContainer.cs (hunk 1 above)
Comment (normalid-awa) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:79]: > This doesn't look good at all and doesn't make sense in the context of how other apps handle this.
> 
> `AllowMultiline` can't be used because it crashes on retrieving the base line height (not allowed on multiline `SpriteText` for some reason).
> 
> But, even if multiline was to be used, TFC will still not handle correctly and will put the following word always on a new line because the spritetext covers the entire horizontal area.
> 
> Basically, TFC needs a lot of work (framework-side) to do things correctly but in the meantime I'd rather not have this here because it looks wrong.

I've set Truncate and MaxWidth at the same time because while the TFC handles the line-break separate by words, it doesn't consider when the word is so long like there are no spaces between the long word, so I set the max width and truncate property to the spite text so that it shows the ellipsis instead of just cut the word.

Comment (smoogipoo) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:79]: Yes, that's not the proper way to handle it. The proper way is `AllowMultiline`, if it worked, or something equivalent implemented inside TFC, i.e., splitting by character if the line is too long.

Try it out in a browser: https://jsfiddle.net/361nLwaq/
image

Comment (normalid-awa) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:79]: > Yes, that's not the proper way to handle it. The proper way is `AllowMultiline`, if it worked, or something equivalent implemented inside TFC, i.e., splitting by character if the line is too long.
> 
> Try it out in a browser: https://jsfiddle.net/361nLwaq/ image

So the only way to make these changes work as expected, is to fix the o!f code?

Comment (smoogipoo) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:79]: Correct. Only wrapping per-word is fine for now because the o!f changes will not be easy to do.

Comment (normalid-awa) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:79]: Is it okay to leave a to-do in here? For the future when the o!f fix that.

After the change: 
``` C#
Children = new Drawable[]
{
    background = new Box
    {
        RelativeSizeAxes = Axes.Both,
        Alpha = 0.9f,
    },
    text = new TextFlowContainer(f =>
    {
        f.Font = OsuFont.GetFont(weight: FontWeight.Regular);
        //TODO: when the o!f supports the AllowMultiline then this SpriteText can use that property to wrap the long one word
        //f.AllowMultiline = true;
        //f.MaxWidth = max_width;
    })
    {
        Margin = new MarginPadding(5),
        AutoSizeAxes = Axes.Both,
        MaximumSize = new Vector2(max_width, float.PositiveInfinity),
    }
};
```

Comment (smoogipoo) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs:79]: It's not going to be/can't be a `SpriteText` implementation, for one of the reasons I mentioned above. It'll be done automatically by TFC.

Comment (normalid-awa) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs]: The latest commit should be okay. Removed the problem code

Comment (smoogipoo) [reply osu.Game/Graphics/Cursor/OsuTooltipContainer.cs]: o!f issue: https://github.com/ppy/osu-framework/issues/5007

**Review by smoogipoo — DISMISSED** [review]

**Comment by smoogipoo** [comment]
Requesting one more review to confirm I haven't gone off the rails with my suggestions (especially maximum size of 500).

**Comment by peppy** [comment]
@smoogipoo how are you testing this? is there a test scene i'm missing?

**Comment by normalid-awa** [comment]
> @smoogipoo how are you testing this? is there a test scene i'm missing?

I think there's no test scene to test the tooltips, I think I could probably open a pr to implement that. But I'm not sure how to emulator the hover event.

**Comment by normalid-awa** [comment]
> @smoogipoo how are you testing this? is there a test scene i'm missing?

I've written the unit test now, not sure if I have to open a new pr or just push to this branch

**Comment by peppy** [comment]
@normalid-awa you can commit here

**Comment by normalid-awa** [comment]
> @normalid-awa you can commit here

Unit test was pushed, One of the test was disabled temporarily due to the o!f issue. [ref discussion](https://github.com/ppy/osu/pull/29210#discussion_r1701714516)

**Review by peppy — APPROVED** [review]

**Comment by peppy** [comment]
@normalid-awa please check the changes i made to your tests to improve things slightly. otherwise looks fine!

**Comment by normalid-awa** [comment]
> @normalid-awa please check the changes i made to your tests to improve things slightly. otherwise looks fine!

Oh...., I was a bit confused at that point, the test should be enable but only disabled the problem case. The changes looks good to me!
